winpr/libwinpr/ncrypt/ncrypt_pkcs11.c

```diff
@@ -793,6 +813,117 @@ static SECURITY_STATUS NCryptP11EnumKeys(NCRYPT_PROV_HANDLE hProvider, LPCWSTR p
 	return NTE_NO_MORE_ITEMS;
 }
 
+static SECURITY_STATUS get_piv_container_name(NCryptP11KeyHandle* key, BYTE* piv_tag, BYTE* output,
+                                              size_t output_len)
+{
+	CK_SLOT_INFO slot_info = { 0 };
+	CK_FUNCTION_LIST_PTR p11 = NULL;
+	WCHAR* reader = NULL;
+	SCARDCONTEXT context = 0;
+	SCARDHANDLE card = 0;
+	DWORD proto = 0;
+	const SCARD_IO_REQUEST* pci = NULL;
+	BYTE buf[60] = { 0 };
+	char container_name[PIV_CONTAINER_NAME_LEN + 1] = { 0 };
+	DWORD buf_len = 0;
+	SECURITY_STATUS ret = NTE_BAD_KEY;
+	WinPrAsn1Decoder dec = { 0 };
+	WinPrAsn1Decoder dec2 = { 0 };
+	size_t len = 0;
+	BYTE tag = 0;
+	BYTE* p = NULL;
+	wStream s = { 0 };
+
+	WINPR_ASSERT(key);
+	WINPR_ASSERT(piv_tag);
+
+	WINPR_ASSERT(key->provider);
+	p11 = key->provider->p11;
+	WINPR_ASSERT(p11);
+
+	/* Get the reader the card is in */
+	WINPR_ASSERT(p11->C_GetSlotInfo);
+	if (p11->C_GetSlotInfo(key->slotId, &slot_info) != CKR_OK)
+		return NTE_BAD_KEY;
+
+	fix_padded_string((char*)slot_info.slotDescription, sizeof(slot_info.slotDescription));
+	if (ConvertToUnicode(
+	        CP_UTF8, 0, (char*)slot_info.slotDescription,
+	        strnlen((char*)slot_info.slotDescription, sizeof(slot_info.slotDescription)), &reader,
+	        0) < 0)
+		return NTE_NO_MEMORY;
+
+	if (SCardEstablishContext(SCARD_SCOPE_USER, NULL, NULL, &context) != SCARD_S_SUCCESS)
+		return NTE_BAD_KEY;
+
+	if (SCardConnectW(context, reader, SCARD_SHARE_SHARED, SCARD_PROTOCOL_Tx, &card, &proto) !=
+	    SCARD_S_SUCCESS)
+		goto out;
+	pci = (proto == SCARD_PROTOCOL_T0) ? SCARD_PCI_T0 : SCARD_PCI_T1;
+
+	buf_len = 2;
+	if (SCardTransmit(card, pci, APDU_PIV_SELECT_AID, sizeof(APDU_PIV_SELECT_AID), NULL, buf,
+	                  &buf_len) != SCARD_S_SUCCESS)
+		goto out;
+	if ((buf[0] != 0x90 || buf[1] != 0) && buf[0] != 0x61)
```

Comment:
This is not entirely correct, if the the result is `61XX` the spec says that you're supposed to issue a `GET_RESPONSE` command like in https://github.com/hardening/FreeRDP/blob/gw_smartcard/winpr/libwinpr/ncrypt/ncrypt_pkcs11.c#L870 and SW2 contains the remaining bytes